cpp/src/arrow/c/abi.h

```diff
@@ -228,6 +228,65 @@ struct ArrowDeviceArrayStream {
 
 #endif  // ARROW_C_DEVICE_STREAM_INTERFACE
 
+#ifndef ARROW_C_ASYNC_STREAM_INTERFACE
+#define ARROW_C_ASYNC_STREAM_INTERFACE
+
+// Similar to ArrowDeviceArrayStream, except designed for an asynchronous
+// style of interaction. While ArrowDeviceArrayStream provides producer
+// defined callbacks, this is intended to be created by the consumer instead.
+// The consumer passes this handler to the producer, which in turn uses the
+// callbacks to inform the consumer of events in the stream.
```

Comment:
What this doesn't state is whether the producer must serialize calls to these callbacks, or if, for example, it is allowed to invoke `on_next` from several threads at once.

Reply:
Oops, I just saw @westonpace 's question below. Sorry :-)